packages/babel-plugin-proposal-partial-application/src/index.js

```diff
@@ -0,0 +1,192 @@
+      if (!t.isArgumentPlaceholder(arg) && !t.isNumericLiteral(arg)) {
+        argId = scope.generateUidIdentifier("_param");
+        scope.push({ id: argId });
+        if (t.isSpreadElement(arg)) {
+          newNode = t.assignmentExpression(
+            "=",
+            t.cloneNode(argId),
+            t.arrayExpression([arg]),
+          );
+        } else {
+          newNode = t.assignmentExpression("=", t.cloneNode(argId), arg);
+        }
+        acc.push(newNode);
+      }
+      return acc;
+    }, []);
+  }
+
+  /**
+   * Unwraps all of the arguments in a CallExpression
+   * and replaces ArgumentPlaceholder type with Identifier
+   * and gives them a unique name.
+   * @param {Object} node
+   * @param {Object} scope
+   * @returns {Array<Expression>} the arguments
+   */
+  function unwrapAllArguments(node, scope) {
+    const clone = t.cloneNode(node);
+
+    return clone.arguments.map(arg => {
+      if (t.isArgumentPlaceholder(arg)) {
+        return Object.assign({}, arg, {
+          type: "Identifier",
+          name: scope.generateUid("_argPlaceholder"),
+        });
+      }
+      return arg;
+    });
+  }
+
+  /**
+   * It replaces the values of non-placeholder args in allArgs
+   * @param {Array<Node>} nonPlaceholderArgs that has no placeholder in them
+   * @param {Array<Node>} args
+   */
+  function mapNonPlaceholderToLVal(nonPlaceholderArgs, allArgsList) {
+    allArgsList.forEach(arg => {
+      nonPlaceholderArgs.forEach(pArg => {
+        if (
+          t.isSpreadElement(arg) &&
+          !t.isIdentifier(pArg.right) &&
+          pArg.right.elements[0].argument.name === arg.argument.name
+        ) {
+          arg.argument.name = pArg.left.name;
+        } else if (!t.isNumericLiteral(arg) && pArg.right.name === arg.name) {
+          arg.name = pArg.left.name;
+        }
+      });
+    });
+    return allArgsList;
+  }
+
+  /**
+   * Takes the full list of arguments and extracts placeholders from it
+   * @param {Array<Argument>} allArgsList full list of arguments
+   * @returns {Array<ArgumentPlaceholder>} cloneList
+   */
+  function placeholderLVal(allArgsList) {
+    const cloneList = [];
+    allArgsList.forEach(item => {
+      if (item.name && item.name.includes("_argPlaceholder")) {
+        cloneList.push(item);
+      }
+    });
+    return cloneList;
+  }
+
+  return {
+    name: "proposal-partial-application",
+    inherits: syntaxPartialApplication,
+
+    visitor: {
+      CallExpression(path) {
+        if (!hasArgumentPlaceholder(path.node)) {
+          return;
+        }
+        const { node, scope } = path;
+        const functionLVal = path.scope.generateUidIdentifierBasedOnNode(
+          node.callee,
+        );
+
+        const nonPlaceholderArgs = unwrapArguments(node, scope);
+        const allArgs = unwrapAllArguments(node, scope);
+        const finalArgs = mapNonPlaceholderToLVal(nonPlaceholderArgs, allArgs);
+        const placeholderVals = placeholderLVal(allArgs);
```

Comment:
While this works, I think that it can be made better: relying on an automatically generated name in `placeholderLVal` doesn't feel 100% ok, and the two nested loops in `mapNonPlaceholderToLVal` can be avoided.

You could merge those functions like this:
```
const argsInitializers = unwrapArguments(node, scope);
const [placeholdersParams, args] = replacePlaceholders(node, scope); 
```

Where `unwrapsArguments` returns the `_param1 = foo, _param2 = bar` part from `fn(foo, ?, bar, ?)` _and_ replaces the parameters: the call becomes `fn(_param1, ?, _param2, ?)`. It could be implemented roughly like this:

```js
function unwrapArguments(node, scope) {
  var init = [];
  for each (arg in node.arguments with index i) {
    if (arg is neither a placeholder nor an immutable value) {
      var id = generate param id;
      init.push("${id} = ${arg}")
      node.arguments[i] = t.cloneNode(id);
    }
  }
  return init;
}
```

`replacePlaceholders` would instead transform `fn(_param1, ?, _param2, ?)` to `fn(_param1, _placeholder1, _param2, _placeholder2)`, and additionally return an array `[_placeholder1, _placeholder2]` to be used to create the arguments of the new function:

```js
function replacePlaceholders(node, scope) {
  var placeholders = [];
  var args = [];
  for each (arg in node.arguments) {
    if (arg is a placeholder) {
      var id = generate placeholder id;
      placeholders.push(id);
      args.push(t.cloneNode(id));
    } else {
      args.push(arg);
    }
  }
  return [placeholders, args];
}
```

At the end, you would need to create the result like this:
```js
functionLVal = FN, ...argsInitializers, function NAME(...placeholdersParams) { return functionLVal(...args) }
```

Reply:
This is a really good idea, but I cannot figure out how to handle the `spreadElement`. My implementation of the two functions:

```js
function unwrapArguments(node, scope) {
    const init = [];
    for (let i = 0; i < node.arguments.length; i++) {
      if (
        !t.isArgumentPlaceholder(node.arguments[i]) &&
        !t.isImmutable(node.arguments[i])
      ) {
        const id = scope.generateUidIdentifierBasedOnNode(
          node.arguments[i],
          "param",
        );
        scope.push({ id });
        if (t.isSpreadElement(node.arguments[i])) {
          init.push(
            t.assignmentExpression(
              "=",
              t.cloneNode(id),
              t.arrayExpression([node.arguments[i]]),
            ),
          );
        } else {
          init.push(
            t.assignmentExpression("=", t.cloneNode(id), node.arguments[i]),
          );
        }
        node.arguments[i] = t.cloneNode(id);
      }
    }
    return init;
  }

  function replacePlaceholders(node, scope) {
    const placeholders = [];
    const args = [];

    node.arguments.forEach(arg => {
      if (t.isArgumentPlaceholder(arg)) {
        const id = scope.generateUid("_argPlaceholder");
        placeholders.push(t.identifier(id));
        args.push(t.cloneNode(t.identifier(id)));
      } else {
        args.push(arg);
      }
    });
    return [placeholders, args];
  }
```
In `unwrapArguments`, I can easily created the assignment for the `SpreadElement`, but when we are pushing the arguments into the `args` in `replacePlaceholders`, there is no information about the type of the argument to create the following form:
```js
const q = (_foo = foo, _a = a, _param = [...b], function foo(_argPlaceholder) {
   return _foo(_a, ..._param, _argPlaceholder);
});
```
and I would always endup with:

```js
const q = (_foo = foo, _a = a, _param = [...b], function foo(_argPlaceholder) {
   return _foo(_a, _param, _argPlaceholder);
});
```

Reply:
Maybe we could somehow in `unwrapArguments` send the needed information along with `init`?

Reply:
Instead of `node.arguments[i] = t.cloneNode(id);`, for spread element you need to do `node.arguments[i].argument = t.cloneNode(id);`.